### PR TITLE
Feature typos detection

### PR DESCRIPTION
**Pull request by flbulgarelli**
This PR adds supports for detecting typos after when following expectations are used: 

*  `Uses` 
* `Declares` and its family: 
    * `Declares.+`
    *  `Delegates`
    * `SubordinatesDeclarationsTo`

The main intention of this feature is to add _did-you-mean_-like capabilities to platforms that use mulang, since I empirically observed they are too-common mistakes that usually too-much time to be detected and understood. 

This feature also introduces the proto-concept of an `Anayzer`'s intermediate state, which may be called `AnalysisContext` in the future, which may be passed and returned by other analyzing steps. 

Currently, only a list of `QueryResult` has been added to the code base. It is consumed by the `SmellsAnalyzer`, which has been refactored in order to be able to produce context-dependent `SmelInstance`s instead of just `Smell`s.

Finally, this PR suggest - but does not fully reify - the concept of native/raw `Detection`s, which are not the result of an inspection combined with `locate` or `detect`, but are first-class generators of identifiers. 

The typos detection has been implemented using [Damerau modification](https://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance) of the Levenshtein distance. I had originally planned to use the latter, but after some reading I understood that for this class of spell checking the first is better, since it considers a single transposition as a distance-1 instead of distance-2.

### CODE REVIEW

File: stack.yaml.ghcjs [hunk 1]
```diff
@@ -15,5 +15,7 @@ extra-deps:
   commit: e65815417ffd464bfb6d5444f7ff1ad4778a9417
 - github: julian-berbel/language-python
   commit: cbdf9b33cc3c30e66c8455165ed1c7d70a300a3e
+- github: mrkkrp/text-metrics
+  commit: 4b2b8c4613d4e948851bc81d823f9e158aadd6c3
```
Comment (flbulgarelli) [inline stack.yaml.ghcjs:19]: It looks like the original implementation of this package was C-based and it is the probable cause of not being part of the ghcjs distro. Nevertheless, I had to manually add it it order JS runtime to work.

**Comment by flbulgarelli** [comment]
PS: I'd really like some feedback about English and Portuguese translations @julian-berbel @Gustrucco @lauramangifesta @ivanafeldfeber :smile:

File: src/Language/Mulang/Analyzer/SmellsAnalyzer.hs [hunk 2]
```diff
@@ -51,56 +82,102 @@ allSmells = [
...
 withLanguage :: (DomainLanguage -> Inspection) -> Detection
-withLanguage inspection language = locate' (inspection language)
+withLanguage inspection context = locate' (inspection (snd context))
+
+raw :: (Expression -> [Identifier]) -> Detection
+raw detection _ = detection
 
-exectationFor :: Smell -> Identifier -> Expectation
-exectationFor smell identifier = Expectation identifier smell
+exectationFor :: SmellInstance -> Identifier -> Expectation
```
Comment (julian-berbel) [inline src/Language/Mulang/Analyzer/SmellsAnalyzer.hs:178]: typo! :smile:

**Comment by julian-berbel** [comment]
> PS: I'd really like some feedback about English and Portuguese translations @julian-berbel @Gustrucco @lauramangifesta @ivanafeldfeber

Regarding this, the translation is good, though the original text seems a little stiff :sweat_smile:

How about something like "Solution must declare bar, but declares baz. Perhaps you meant bar?"

File: gem/lib/locales/pt.yml [hunk 3]
```diff
@@ -109,6 +109,7 @@ pt:
       DoesTypeTest: '%{binding}  faz comparações contra strings'
       HasAssignmentReturn: '%{binding}  retorna o resultado de uma tarefa'
       HasCodeDuplication: '%{binding}  tem código repetido'
+      HasDeclarationTypos_named: 'A solução deve declarar %{target}, mas declarar %{binding}. Você poderia ter cometido um erro de digitação?'
```
Comment (ivanafeldfeber) [inline gem/lib/locales/pt.yml:112]: A solução deve declarar %{target}, mas declara %{binding}. Você poderia ter cometido um erro de digitação?'

File: gem/lib/locales/pt.yml [hunk 4]
```diff
@@ -109,6 +109,7 @@ pt:
...
       HasEmptyIfBranches: '%{binding} tem vazio <i>%{keyword_if}</i> ramos'
```
Comment (ivanafeldfeber) [inline gem/lib/locales/pt.yml:113]: '%{binding} tem uma ramificaçao <i>%{keyword_if}</i> vazia'

File: gem/lib/locales/pt.yml [hunk 5]
```diff
@@ -126,6 +127,7 @@ pt:
       HasTooShortBindings: '%{binding}  é um identificador muito curto'
       HasTooShortIdentifiers: '%{binding} é um identificador muito curto'
       HasUnreachableCode: '%{binding} tem código inacessível'
+      HasUsageTypos_named: 'A solução deve usar %{target}, mas use %{binding}. Você poderia ter cometido um erro de digitação?'
```
Comment (ivanafeldfeber) [inline gem/lib/locales/pt.yml:130]: 'A solução deve usar %{target}, mas usa %{binding}. Você poderia ter cometido um erro de digitação?'

**Pull request merged**